### DISCUSSION TITLE
Does Duckdb support connecting to the same in-memory instance through multiple database connections within a process?

### DISCUSSION THREAD

**Discussion by robotslacker** [duckdb/duckdb#13078] [Q&A] [answered]
Try to create multiple database connections in a process and connect to the same database file. This is successful and there is no problem.

```
        String  connectURL = "jdbc:duckdb:aa.duckdb";
        int  MAX_THREADS = 5;

        // Create 100 threads
        Thread[] threads = new Thread[MAX_THREADS];

        Class.forName("org.duckdb.DuckDBDriver");

        Connection conn = DriverManager.getConnection(connectURL, "", "");
        conn.setAutoCommit(false);
        conn.createStatement().execute("Create TABLE lotsOfConnection (id int)");
        conn.commit();

        Sleeper.sleep(3000);

        for (int i = 0; i < threads.length; i++) {
            int finalI = i;
            threads[i] = new Thread(() -> {
                try {
                    Connection conn2 = DriverManager.getConnection(connectURL, "", "");
                    conn2.setAutoCommit(false);
                    Statement stmt = conn2.createStatement();
                    stmt.execute("insert into lotsOfConnection values(" + finalI + ")");
                    conn2.commit();
                    conn2.close();
                } catch (SQLException se) {
                    se.printStackTrace();
                }
            });
        }

        for (Thread thread : threads) {
            thread.start();
        }

        // Waiting for all threads completed
        for (Thread thread : threads) {
            try {
                thread.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        int expectedSumValue = 0;
        for (int i = 0; i < threads.length; i++) {
            expectedSumValue = expectedSumValue + i;
        }
        ResultSet rs = conn.createStatement().executeQuery("SELECT COUNT(*),SUM(id) from lotsOfConnection");
        while (rs.next()) {
            System.out.println(rs.getInt(1));
            System.out.println(rs.getInt(2));
        }
        rs.close();

```

But try to create multiple database connections in a process and connect to the same memory instance. It seems that each connection sees its own memory copy, not the same database.
``` 
        String  connectURL = "jdbc:duckdb::memory:";
        .....
```

Does the memory database support naming? Like :memory:aa, :memory:bb

**Comment by Tishj** [comment, marked as answer]
It does, but it might depend on which client you're using `:memory:custom_name` will work in those clients that do (such as python, odbc and likely a lot more)

Looking at [the source](https://github.com/duckdb/duckdb-java/blob/main/src%2Fjni%2Fduckdb_java.cpp#L419) this should be possible in jdbc as well, as the DBInstanceCache is what enables this functionality

**Reply by robotslacker** [reply]
Yes. It does. It works! Thanks a lot.
Maybe we should add something in document.

**Reply by mattiapiccinetti** [reply]
@robotslacker Could you share a working snippet? It didn’t work for me (snippet below). I have one process that creates tables and another that writes records, but the latter can't see the tables just created. 
I used SingleConnectionDataSource as a workaround, but then I ran into the issues described [here](https://github.com/duckdb/duckdb/issues/11373)

```
@Bean
public DataSource dataSource() {
    var dataSource = new DriverManagerDataSource();
    dataSource.setDriverClassName("org.duckdb.DuckDBDriver");
    dataSource.setUrl("jdbc:duckdb::memory:foo_db");
    return dataSource;
}
```

**Comment by robotslacker** [comment]
Connection&nbsp; backendSysConnection = DriverManager.getConnection(""jdbc:duckdb:xxxx");
# leave this backend connection open, don't close it!!!
# in 1.3.0.1, new feature in duckdb_java will add a new parameter to let you keep database open even you close all connecton.




Thread 1：
Connection conn = backendSysConnection.duplicate();
// create table ...


Thread2:
Connection conn = backendSysConnection.duplicate();

// query table ....


Hope this can help you.






RobotSlacker
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "duckdb/duckdb"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年6月9日(星期一) 晚上11:39
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [duckdb/duckdb] Does Duckdb support connecting to the same in-memory instance through multiple database connections within a process? (Discussion #13078)





 
@robotslacker Could you share a working snippet? It didn’t work for me. I have one process that creates tables and another that writes records, but the latter can't see the tables just created.
 I used SingleConnectionDataSource as a workaround, but then I ran into the issues described here
 @Bean public DataSource dataSource() {     var dataSource = new DriverManagerDataSource();     dataSource.setDriverClassName("org.duckdb.DuckDBDriver");     dataSource.setUrl("jdbc:duckdb::memory:foo_db");     return dataSource; }

**Comment by robotslacker** [comment]
Correct：


Connection&nbsp; backendSysConnection = DriverManager.getConnection("jdbc:duckdb:xxxx");
#&nbsp;leave this backend connection open, don't close it!!!
# in 1.3.0.1, new feature in duckdb_java will add a new parameter to let you keep database open even you close all connecton.




Thread 1：
Connection conn = backendSysConnection.duplicate(); or = DriverManager.getConnection("jdbc:duckdb:xxxx")
// create table ...
// commit;&nbsp; &nbsp; &nbsp;&nbsp;


Thread2:
Connection conn = backendSysConnection.duplicate();&nbsp; or = DriverManager.getConnection("jdbc:duckdb:xxxx")

// query table ....


Hope this can help you.





RobotSlacker
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "RobotSlacker"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年6月10日(星期二) 中午1:38
***@***.******@***.***&gt;;
***@***.***&gt;;
主题:&nbsp;回复： [duckdb/duckdb] Does Duckdb support connecting to the same in-memory instance through multiple database connections within a process? (Discussion #13078)



Connection&nbsp; backendSysConnection = DriverManager.getConnection(""jdbc:duckdb:xxxx");
# leave this backend connection open, don't close it!!!
# in 1.3.0.1, new feature in duckdb_java will add a new parameter to let you keep database open even you close all connecton.




Thread 1：
Connection conn = backendSysConnection.duplicate();
// create table ...


Thread2:
Connection conn = backendSysConnection.duplicate();

// query table ....


Hope this can help you.






RobotSlacker
***@***.***



&nbsp;




------------------ 原始邮件 ------------------
发件人:                                                                                                                        "duckdb/duckdb"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年6月9日(星期一) 晚上11:39
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [duckdb/duckdb] Does Duckdb support connecting to the same in-memory instance through multiple database connections within a process? (Discussion #13078)





 
@robotslacker Could you share a working snippet? It didn’t work for me. I have one process that creates tables and another that writes records, but the latter can't see the tables just created.
 I used SingleConnectionDataSource as a workaround, but then I ran into the issues described here
 @Bean public DataSource dataSource() {     var dataSource = new DriverManagerDataSource();     dataSource.setDriverClassName("org.duckdb.DuckDBDriver");     dataSource.setUrl("jdbc:duckdb::memory:foo_db");     return dataSource; }

**Comment by mattiapiccinetti** [comment]
Thanks @robotslacker! I extended the SingleConnectionDataSource class, overriding the getConnection method to return a duplicated DuckDBConnection (like you suggested), and it worked! I needed a DataSource object, so I created a custom one
